### PR TITLE
REGRESSION (r289032): rotate animation doesn't interpolate between 0 and 1turn without forced 50% https://bugs.webkit.org/show_bug.cgi?id=239906

### PR DESCRIPTION
**Pull request by mrobinson**
#### 43e80183650b2053fb865f81218eb3c94724f42c
<pre>
REGRESSION (r289032): rotate animation doesn&apos;t interpolate between 0 and 1turn without forced 50% <a href="https://bugs.webkit.org/show_bug.cgi?id=239906">https://bugs.webkit.org/show_bug.cgi?id=239906</a>

Reviewed by Simon Fraser.

When using CoreAnimation non-matrix animations to animate rotations,
CoreAnimation will use the shortest direction between two rotation
angles. This means that a rotation from 0 to 360 will not rotate at all.
This is different from how CSS works, where it expects the animation to
do a full turn. In order to avoid problems with this difference, when an
animation includes larger angles ( &gt; 180 degrees), fall back to software
animation.

No new tests. It is difficult to make a test for this because
when pausing animations the software path is used.

* LayoutTests/animations/3d/full-rotation-animation-expected.html: Added.
* LayoutTests/animations/3d/full-rotation-animation.html: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::transformationAnimationValueAt):
(WebCore::hasBigRotationAngle):
(WebCore::GraphicsLayerCA::createTransformAnimationsFromKeyframes):
(WebCore::GraphicsLayerCA::setTransformAnimationEndpoints):
* Source/WebCore/platform/graphics/transforms/TransformOperations.h:
(WebCore::SharedPrimitivesPrefix::primitives const):
(WebCore::SharedPrimitivesPrefix::primitives): Deleted.

Canonical link: <a href="https://commits.webkit.org/250920@main">https://commits.webkit.org/250920@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@294752">https://svn.webkit.org/repository/webkit/trunk@294752</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by graouts** [comment]
I can't think of a testing strategy here either given it's really about interpolation behavior across time for a rotation. I've tagged @smfr to review, maybe he'll have a bright idea! If not, I think this warrants an exception to land without new test coverage.

**Comment by smfr** [comment]
> #### [2975ca4](https://github.com/WebKit/WebKit/commit/2975ca4c4ec16bf21ed66c9d1d167f1c384c34f1)
> REGRESSION (r289032): rotate animation doesn't interpolate between 0 and 1turn without forced 50%
> https://bugs.webkit.org/show_bug.cgi?id=239906
> 
> Reviewed by NOBODY (OOPS!).
> 
> No new tests. It is difficult to make a test for this because
> when pausing animations the software path is used.

Can't you make an animation with a 10,000yr duration and a delay of 5,000yr to effectively pause it in the middle?

**Comment by graouts** [comment]
> > #### [2975ca4](https://github.com/WebKit/WebKit/commit/2975ca4c4ec16bf21ed66c9d1d167f1c384c34f1)
> > REGRESSION (r289032): rotate animation doesn't interpolate between 0 and 1turn without forced 50%
> > https://bugs.webkit.org/show_bug.cgi?id=239906
> > Reviewed by NOBODY (OOPS!).
> > No new tests. It is difficult to make a test for this because
> > when pausing animations the software path is used.
> 
> Can't you make an animation with a 10,000yr duration and a delay of 5,000yr to effectively pause it in the middle?

Right, of course. This ought to work, but I always worry about pixel diff for ref tests with rotations. Maybe if it's at 180deg it'll be fine.

**Comment by mrobinson** [comment]
> Can't you make an animation with a 10,000yr duration and a delay of 5,000yr to effectively pause it in the middle?

Okay. I've uploaded a new version of this PR with a test that uses the strategy you describe here. I further tried to avoid pixel differences by making it so the correct state was a 90 degree rotation around the y axis -- making the rotated div completely invisible. I've confirmed that this test fails in STP and passes with my changes.

File: LayoutTests/animations/3d/full-rotation-animation.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  </div>
+</div>
+
+<script>
+    let target = document.querySelector('.box');
+    getComputedStyle(target).getPropertyValue("transform");
+    target.style.animationName = 'animation';
+    target.style.animationTimingFunction = 'linear';
+    target.style.animationDuration = '10000000000s';
+    target.style.animationDelay = '-2500000000s';
```
Comment (graouts) [inline LayoutTests/animations/3d/full-rotation-animation.html]: Why are those required vs. just setting them in the `<style>` element above?

Comment (mrobinson) [reply LayoutTests/animations/3d/full-rotation-animation.html]: Good point. I've moved these into the `<style>` element.

**Comment by smfr** [comment]
> #### [5653477](https://github.com/WebKit/WebKit/commit/5653477e2d8d17618bb0f87de3a131f182dad4f4)
> REGRESSION (r289032): rotate animation doesn't interpolate between 0 and 1turn without forced 50%
> https://bugs.webkit.org/show_bug.cgi?id=239906
> 
> Reviewed by NOBODY (OOPS!).
> 
> No new tests. It is difficult to make a test for this because
> when pausing animations the software path is used.

There's no explanation here of what the changes are and why you had to make them.

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 2]
```diff
@@ -3477,6 +3477,30 @@ bool GraphicsLayerCA::appendToUncommittedAnimations(const KeyframeValueList& val
...
+static const TransformOperations& transformationAnimationValueAt(const KeyframeValueList& valueList, unsigned i)
+{
+    return static_cast<const TransformAnimationValue&>(valueList.at(i)).value();
+}
+
+static bool hasBigRotationAngle(const KeyframeValueList& valueList, const SharedPrimitivesPrefix& prefix)
+{
+    const auto& primitives = prefix.primitives();
+    for (unsigned animationIndex = 0; animationIndex < primitives.size(); ++animationIndex) {
+        TransformOperation::OperationType type = primitives[animationIndex];
```
Comment (smfr) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: auto

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp [hunk 3]
```diff
@@ -3477,6 +3477,30 @@ bool GraphicsLayerCA::appendToUncommittedAnimations(const KeyframeValueList& val
...
+        if (type != TransformOperation::ROTATE && type != TransformOperation::ROTATE_3D)
+            continue;
+        for (size_t i = 1; i < valueList.size(); ++i) {
+            auto prevOperation = static_cast<const RotateTransformOperation*>(transformationAnimationValueAt(valueList, i - 1).at(animationIndex));
```
Comment (smfr) [inline Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: This should use downcast<> (as should the next line).

It's hard to grok the relationship between valueList and prefix, and I can't tell how you know that the previous operation is a rotate operation.

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp (hunk 2 above)
Comment (mrobinson) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: Fixed!

File: Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp (hunk 3 above)
Comment (mrobinson) [reply Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp]: I've switched this to use `downcast<>` and also have added a comment giving a bit more background about why the prefix used here.

**Comment by mrobinson** [comment]
> There's no explanation here of what the changes are and why you had to make them.

Sorry. This is included now.

**Comment by mrobinson** [comment]
The failing test is a crash in `LayoutTests/http/tests/media/modern-media-controls/time-control/less-than-10-minutes.html` which seems unrelated to this change.

**Review by smfr — APPROVED** [review]
